nvme_gw_persistence.py

```diff
@@ -0,0 +1,443 @@
+        self._delete_key(key)
+
+    def _restore_namespaces(self, omap_dict):
+        """Restores a namespace from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.NAMESPACE_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                self.logger.info(f"Restoring namespace: {args}")
+                try:
+                    return_string = self.spdk_rpc.nvmf.nvmf_subsystem_add_ns(
+                        self.spdk_rpc_client,
+                        args["subsystem_nqn"],
+                        args["bdev_name"],
+                    )
+                    self.logger.info(f"Added NSID to subystem: {return_string}")
+                except Exception as ex:
+                    self.logger.error(
+                        f"nvmf_subsystem_add_ns failed with: \n {ex}")
+                    raise
+
+    def add_subsystem(self, subsystem_nqn: str, val: str):
+        """Adds a subsystem to the persistent config."""
+        key = self.SUBSYSTEM_PREFIX + subsystem_nqn
+        self._write_key(key, val)
+
+    def delete_subsystem(self, subsystem_nqn: str):
+        """Deletes a subsystem from the persistent config."""
+        key = self.SUBSYSTEM_PREFIX + subsystem_nqn
+        self._delete_key(key)
+
+    def _restore_subsystems(self, omap_dict):
+        """Restores subsystems from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.SUBSYSTEM_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                self.logger.info(f"Restoring subsystem: {args}")
+                try:
+                    return_string = self.spdk_rpc.nvmf.nvmf_create_subsystem(
+                        self.spdk_rpc_client, args["subsystem_nqn"],
+                        args["serial_number"])
+                    self.logger.info(f"Returned with status: {return_string}")
+                except Exception as ex:
+                    self.logger.error(
+                        f"nvmf_create_subsystem failed with: \n {ex}")
+                    raise
+
+    def add_host(self, subsystem_nqn: str, host_nqn: str, val: str):
+        """Adds a host to the persistent config."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._write_key(key, val)
+
+    def delete_host(self, subsystem_nqn: str, host_nqn: str):
+        """Deletes a host from the persistent config."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._delete_key(key)
+
+    def _restore_hosts(self, omap_dict):
+        """Restore hosts from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.HOST_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                self.logger.info(f"Restoring host: {args}")
+                try:
+                    self.spdk_rpc.nvmf.nvmf_subsystem_add_host(
```

Comment:
I don't like very much that we are duplication here the spdk_rpc code we have in nvme_gw_server, and that we are mixing these thing together. Ideally the `PersistentConfig` would not need any knowledge about spdk_rpc.

There are many possible solutions to this. E.g. instead of `persistent_config.restore` you could provide `persistent_config.load` that would return the config to  `GWService` and it would talk to the target itself (preferably just reusing the functions it already has).

Or you could have a callback param to your restore functions and the `GWService` would set it to its own corresponding functions.